### PR TITLE
Multiscale ts image computation

### PR DESCRIPTION
**Pull request by adonath**
I now there is still a bit of work to be done here (put the script in a main function etc., change to back `astropy.nddata.utils.extract_array`, docs missing...), however I think its ready for a first review.

### CODE REVIEW

File: gammapy/detect/test_statistics.py [hunk 1]
```diff
@@ -232,30 +357,70 @@ def _ts_value(position, counts, exposure, background, kernel, flux,
...
-            C_1 = cash(counts_slice, background_slice + amplitude * FLUX_FACTOR * model).sum()
+            C_1 = cash(counts_slice, background_slice + FLUX_FACTOR *
+                       amplitude * model).sum()
 
+        # Compute and return TS value
         return C_0 - C_1, amplitude, niter
 
 
+def _amplitude_bounds(counts, background, model):
+    """
```
Comment (cdeil) [inline gammapy/detect/test_statistics.py]: Add one-line description what this function does.

File: gammapy/detect/test_statistics.py [hunk 2]
```diff
@@ -72,30 +93,130 @@ def f_cash(x, counts, background, model):
...
+
+        # Set up PSF and source kernel
+        kernel = multi_gauss_psf_kernel(psf_parameters, BINSZ=BINSZ,
+                                        NEW_BINSZ=BINSZ * factor,
+                                        mode='oversample', x_size=21)
+
+        if scale > 0:
+            from astropy.convolution import convolve
+            sigma = scale / (BINSZ * factor)
+            if morphology == 'Gaussian2D':
```
Comment (cdeil) [inline gammapy/detect/test_statistics.py]: Always end such if -elif ... with an else and throw a `ValueError`.
If you don't the user will mistype the string and then your function will fail in weird hard-to-understand ways later.

File: gammapy/detect/test_statistics.py [hunk 3]
```diff
@@ -72,30 +93,130 @@ def f_cash(x, counts, background, model):
...
         Source template (multiplied with exposure).
     """
     with np.errstate(invalid='ignore', divide='ignore'):
         return cash(counts, background + x * FLUX_FACTOR * model).sum()
 
 
-def compute_ts_map(counts, background, exposure, kernel, flux=None,
-                   method='root', optimizer='Brent', parallel=True,
+def compute_ts_map_multiscale(maps, psf_parameters, scales=[0], downsample='auto',
+                              residual=False, morphology='Gaussian2D', width=0.3,
```
Comment (cdeil) [inline gammapy/detect/test_statistics.py]: The width parameter is not documented.
Probably 0.3 is something specific to your application ... if so this parameter should not have a default.

You should also say what happens to args and kwargs in a sentence ... if those are really needed ... it usually makes for complex code to be flexible in that way.

File: gammapy/detect/test_statistics.py [hunk 4]
```diff
@@ -72,30 +93,130 @@ def f_cash(x, counts, background, model):
...
+        List of `TSMapResult` objects.
+    """
+    BINSZ = abs(maps[0].header['CDELT1'])
+    shape = maps[0].data.shape
+    multiscale_result = []
+
+    for scale, factor in zip(scales, downsample):
+        logging.info('Computing (residual) TS map for scale {0} deg'.format(scale))
+
+        # Sample down and require that scale parameters is at least 5 pix
```
Comment (cdeil) [inline gammapy/detect/test_statistics.py]: This function is quite long ... ideally every function should fit on a screen.

Can you split it up, e.g. extract this part into a `_downsample` helper function?

File: gammapy/detect/test_statistics.py [hunk 5]
```diff
@@ -35,15 +38,33 @@ class TSMapResult(Bunch):
...
-    amplitude : `~numpy.ndarray`
+    amplitude : ndarray
         Estimated best fit flux amplitude map
-    niter : `~numpy.ndarray`
+    niter : ndarray
         Number of iterations map
     runtime : float
         Time needed to compute TS map.
     """
+    def read(self, filename):
```
Comment (cdeil) [inline gammapy/detect/test_statistics.py]: If possible I'd like `read` to always be a `@staticmethod` like `Table.read` in astropy or [EnergyDependentTablePSF.read](https://gammapy.readthedocs.org/en/latest/_modules/gammapy/irf/psf_table.html#EnergyDependentTablePSF.read) in Gammapy.

File: dev/ts_image.py [hunk 6]
```diff
@@ -36,89 +45,48 @@
...
+from gammapy.detect import compute_ts_map_multiscale
 
 
 # Read data
-logging.info('Reading {0}'.format(args.data))
-maps = fits.open(args.data)
-shape = maps[0].data.shape
+logging.info('Reading {0}'.format(args.input))
+maps = fits.open(args.input)
+psf_parameters = json.load(open(args.psf))
```
Comment (cdeil) [inline dev/ts_image.py]: Add logging.info that you're reading the PSF file.

File: dev/ts_image.py [hunk 7]
```diff
@@ -36,89 +45,48 @@
                     help='Input excess model FITS file name')
 parser.add_argument('--threshold', type=float, default=None,
                     help="Minimal required initial (before fitting) TS value, that the fit is done at all.")
-parser.add_argument('--overwrite', action='store_true',
-                    help='Overwrite existing output file?')
+parser.add_argument('--clobber', action='store_true',
```
Comment (cdeil) [inline dev/ts_image.py]: I've opted for consistent use of `overwrite` in `Gammapy` instead of `clobber`.
I was 50:50 on this, but `astropy.table.Table` uses `overwrite` and I think the plan is to use it consistently in Astropy, even if now `astropy.io.fits` uses `clobber`.
OK?

File: dev/ts_image.py [hunk 8]
```diff
@@ -13,22 +12,32 @@
...
+parser.add_argument('--folder', type=str, default='',
+                    help='Output folder name.')
 parser.add_argument('--psf', type=str, default='psf.json', 
                     help='JSON file containing PSF information. ')
-parser.add_argument('--kernel', type=str, default='Gaussian2DKernel', 
-                    help='Which kernel to use for TS calculation.')
-parser.add_argument('--scale', type=float, default=0,
-                    help='Scale of the kernel in deg.')
+parser.add_argument('--morphology', type=str, default='Gaussian2D', 
+                    help="Which source morphology to use for TS calculation. Either 'Gaussian2D' or 'Shell2D'.")
```
Comment (cdeil) [inline dev/ts_image.py]: Wrap line

**Comment by cdeil** [comment]
I've left some stylistic comments inline.

You need to fix the tests:
https://travis-ci.org/gammapy/gammapy/jobs/48174547#L2459

Once that passes and these little things are fixed it's up to you when you want to merge.
I'm working on spectra and `gammapy.data` classes, so no rush from my side.

It would be nice if you moved the script to `gammapy/scripts` and put it in functions so that it's re-usable and shows up in the docs ....

File: gammapy/detect/test_statistics.py [hunk 9]
```diff
@@ -196,7 +319,7 @@ def _ts_value(position, counts, exposure, background, kernel, flux,
         Background map.
     exposure : array
         Exposure map.
-    kernel : `astropy.convolution.Kernel2D`
+    kernel : astropy.convolution.core.Kernel2D
```
Comment (cdeil) [inline gammapy/detect/test_statistics.py]: Why did you change this?
I think the old version is the correct one that creates a HTML link and the new one doesn't?
(see https://gammapy.readthedocs.org/en/latest/development/index.html#where-should-i-import-from)

Comment (adonath) [inline gammapy/detect/test_statistics.py]: I had merge conflicts and just copied the files, I guess that's were this "backward" changes come from. Thanks for the comments so far! I will address them soon and also add an example to gammapy-extra.

File: dev/ts_image.py [hunk 10]
```diff
@@ -13,22 +12,32 @@
 """
 
 # Parse command line arguments
+import warnings
+import os
+
+from astropy.utils.exceptions import AstropyDeprecationWarning
+warnings.filterwarnings('ignore', category=AstropyDeprecationWarning)
```
Comment (hamogu) [inline dev/ts_image.py]: If this is brand-new code, is it not dangerous to ignore astropy depreciation warnings? If you use a feature that's deprecated already, there is a good chance it will be removed in one of the next astropy versions.

Comment (cdeil) [inline dev/ts_image.py]: Agreed ... since we only support Astropy 1.0 (master at the moment) in Gammapy it should be possible to write code that doesn't raise any deprecation warnings.
@adonath If you get deprecation warnings from unrelated parts of Gammapy you don't have to fix them in this PR, but please don't silence them either.

File: gammapy/detect/test_statistics.py [hunk 11]
```diff
@@ -207,10 +330,12 @@ def _ts_value(position, counts, exposure, background, kernel, flux,
...
     """
+    from imageutils import extract_array_2d
+
     # Get data slices
-    counts_slice = extract_array(counts, kernel.shape, position).astype(float)
-    background_slice = extract_array(background, kernel.shape, position).astype(float)
-    exposure_slice = extract_array(exposure, kernel.shape, position).astype(float)
+    counts_slice = extract_array_2d(counts, kernel.shape, position).astype(float)
+    background_slice = extract_array_2d(background, kernel.shape, position).astype(float)
+    exposure_slice = extract_array_2d(exposure, kernel.shape, position).astype(float)
```
Comment (hamogu) [inline gammapy/detect/test_statistics.py]: Can you tell me why you switched back from `astropy.nddata.utils.extract_array` to `imageutils.extract_array_2d`? imageutils has been merged with astropy and will not be developed actively any longer. I'm currently working on a PR to improve `astropy.nddata.utils.extract_array` so please let me know if you found any differences between the two that matter for you.

Comment (adonath) [inline gammapy/detect/test_statistics.py]: This change was not intended. As I had some merge conflicts, I just copied the changed files, but obviously they included some old stuff. Thanks for the hint!

File: dev/ts_image.py (hunk 10 above)
Comment (adonath) [inline dev/ts_image.py]: Yes, there were unrelated deprecation warnings and I included this local fix to have a clean output of the command line tool, when I was testing. Just forgot to remove it...Thanks!

File: gammapy/detect/test_statistics.py [hunk 12]
```diff
@@ -83,11 +129,155 @@ def f_cash(x, counts, background, model):
...
+            sigma = scale / (BINSZ * factor)
+            if morphology == 'Gaussian2D':
+                source_kernel = Gaussian2DKernel(sigma, mode='oversample')
+            elif morphology == 'Shell2D':
+                model = Shell2D(1, 0, 0, sigma, sigma * width)
+                x_size = _round_up_to_odd_integer(2 * sigma * (1 + width)
+                                                  + kernel.shape[0] / 2)
+                source_kernel = Model2DKernel(model, x_size=x_size, mode='oversample')
+            else:
+                raise ValueError('Unknown morphology model.')
```
Comment (cdeil) [inline gammapy/detect/test_statistics.py]: Change to:
```
ValueError('Unknown morphology: {}'.format(morphology))
```

File: gammapy/detect/test_statistics.py [hunk 13]
```diff
@@ -83,11 +129,155 @@ def f_cash(x, counts, background, model):
...
+    ts_max = np.max(ts, axis=2)
+    scale_max = np.zeros(ts.shape[:-1])
+    niter_max = np.zeros(ts.shape[:-1])
+    amplitude_max = np.zeros(ts.shape[:-1])
+
+    for i, scale in enumerate(scales):
+        index = np.where(ts[:, :, i] == ts_max)
+        scale_max[index] = scale
+        niter_max[index] = niter[:, :, i][index]
+        amplitude_max[index] = amplitude[:, :, i][index]
```
Comment (cdeil) [inline gammapy/detect/test_statistics.py]: Add empty line.

File: gammapy/detect/test_statistics.py [hunk 14]
```diff
@@ -83,11 +129,155 @@ def f_cash(x, counts, background, model):
...
+        ts_results = compute_ts_map(maps_['on'], background, maps_['expgammamap'],
+                                    kernel, *args, **kwargs)
+        logging.info('TS map computation took {0:.1f} s \n'.format(ts_results.runtime))
+        ts_results['scale'] = scale
+        ts_results['morphology'] = morphology
+        if downsampled:
+            for name, order in zip(['ts', 'amplitude', 'niter'], [1, 1, 0]):
+                ts_results[name] = upsample_2N(ts_results[name], factor,
+                                               order=order, shape=shape)
+        multiscale_result.append(ts_results)
```
Comment (cdeil) [inline gammapy/detect/test_statistics.py]: Add empty line.

File: gammapy/detect/test_statistics.py [hunk 15]
```diff
@@ -208,54 +398,79 @@ def _ts_value(position, counts, exposure, background, kernel, flux,
...
-        elif method == 'fit scipy':
-            amplitude, niter = _fit_amplitude_scipy(counts_slice, background_slice,
-                                                    model, flux_value)
-        elif method == 'root':
-            amplitude, niter = _root_amplitude(counts_slice, background_slice,
-                                               model, flux_value)
-        if niter > MAX_NITER:
-            logging.warning('Exceeded maximum number of function evaluations!')
-            return np.nan, amplitude, niter
+        raise ValueError('Invalid fitting method.')
```
Comment (cdeil) [inline gammapy/detect/test_statistics.py]: Add empty line.

File: gammapy/detect/test_statistics.py [hunk 16]
```diff
@@ -208,54 +398,79 @@ def _ts_value(position, counts, exposure, background, kernel, flux,
...
+        C_1 = f_cash(amplitude, counts_, background_, model)
+
+    # Compute and return TS value
+    return C_0 - C_1, amplitude * FLUX_FACTOR, niter
 
-        return C_0 - C_1, amplitude, niter
+
+def _amplitude_bounds(counts, background, model):
+    """
+    Compute bounds for     the root of `f_cash_root`.
```
Comment (cdeil) [inline gammapy/detect/test_statistics.py]: `for     the` -> `for the`

File: gammapy/image/utils.py [hunk 17]
```diff
@@ -375,6 +376,47 @@ def coordinates(image, world=True, lon_sym=True, radians=False, system=None):
     return lon, lat
 
 
+def detect_peaks(image, threshold, mask=None, world=True):
```
Comment (cdeil) [inline gammapy/image/utils.py]: Maybe wait with this PR until the photutils version is merged in `photutils` and then use it here?

File: gammapy/scripts/ts_image.py [hunk 18]
```diff
@@ -0,0 +1,89 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+from ..utils.scripts import get_parser
+import os
```
Comment (cdeil) [inline gammapy/scripts/ts_image.py]: Always sort imports:
- Python stdlib
- numpy
- scipy
- Astropy
- Gammapy

File: gammapy/scripts/ts_image.py [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+    parser.add_argument('--threshold', type=float, default=None,
+                        help="Minimal required initial (before fitting) TS value,"
+                        " that the fit is done at all.")
+    parser.add_argument('--overwrite', action='store_true',
+                        help='Overwrite output files.')
+    args = parser.parse_args()
+    ts_image(**vars(args))
+
+
+def ts_image(input, psf, model, scales, downsample, residual, morphology,
```
Comment (cdeil) [inline gammapy/scripts/ts_image.py]: `input` is a Python keyword ... you shouldn't use it for variable names.
Maybe use `data` here or `data_filename`?

**Comment by cdeil** [comment]
I've left some cosmetic inline comments ... this looks good to me know, but tests should be added and if possible the peak finder from `photutils` should be used.

Feel free to merge this yourself next week!
